ext/curl/share.c

```diff
@@ -134,6 +136,131 @@ PHP_FUNCTION(curl_share_strerror)
 }
 /* }}} */
 
+/**
+ * Initialize a persistent curl share handle with the given share options, reusing an existing one if found.
+ *
+ * Throws an exception if the share options are invalid.
+ */
+PHP_FUNCTION(curl_share_init_persistent)
+{
+	zval *share_opts = NULL, *entry = NULL;
+	zend_ulong persistent_id = 0;
+
+	php_curlsh *sh;
+
+	CURLSHcode error;
+
+	ZEND_PARSE_PARAMETERS_START(1, 1)
+		Z_PARAM_ARRAY_EX(share_opts, 0, 1)
+	ZEND_PARSE_PARAMETERS_END();
+
+	object_init_ex(return_value, curl_share_persistent_ce);
```

Comment:
As far as I see, it would make sense to postpone the creation of the returned object to the latest possible time so that the error cases before it don't have to deal with freeing.